File: packages/java/src/main/java/ai/alumnium/Alumni.java
package ai.alumnium;

import ai.alumnium.accessibility.BaseAccessibilityTree;
import ai.alumnium.client.Data;
import ai.alumnium.client.FindElementResult;
import ai.alumnium.client.HttpClient;
import ai.alumnium.driver.AppiumDriver;
import ai.alumnium.driver.BaseDriver;
import ai.alumnium.driver.PlaywrightDriver;
import ai.alumnium.driver.SeleniumDriver;
import ai.alumnium.result.DoResult;
import ai.alumnium.result.DoStep;
import ai.alumnium.tool.BaseTool;
import ai.alumnium.tool.ToolToSchemaConverter;
import ai.alumnium.util.Retry;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Public entry point for the Java client.
 *
 * <p>Driver selection uses exhaustive {@code switch} pattern matching on the supplied native driver
 * object. Supported driver types:
 *
 * <ul>
 *   <li>{@link io.appium.java_client.AppiumDriver} &rarr; {@link AppiumDriver}
 *   <li>{@link com.microsoft.playwright.Page} &rarr; {@link PlaywrightDriver}
 *   <li>{@link org.openqa.selenium.WebDriver} &rarr; {@link SeleniumDriver}
 * </ul>
 */
public final class Alumni implements AutoCloseable {

  private static final Logger LOG = LoggerFactory.getLogger(Alumni.class);

  private final BaseDriver driver;
  private final Model model;
  private final HttpClient client;
  private final Cache cache;
  private final Map<String, Class<? extends BaseTool>> tools;
  private final boolean changeAnalysis;

  public Alumni(Object driver) {
    this(driver, new Options());
  }

  public Alumni(Object driver, Options options) {
    Options opts = options == null ? new Options() : options;
    boolean planner = opts.planner() != null ? opts.planner() : Config.PLANNER;
    this.changeAnalysis =
        opts.changeAnalysis() != null ? opts.changeAnalysis() : Config.CHANGE_ANALYSIS;
    Set<String> excludeAttributes =
        opts.excludeAttributes() != null ? opts.excludeAttributes() : Config.EXCLUDE_ATTRIBUTES;
    this.driver = wrapDriver(driver);

    Map<String, Class<? extends BaseTool>> builtTools = new LinkedHashMap<>();
    for (Class<? extends BaseTool> tool : this.driver.supportedTools()) {
      builtTools.put(tool.getSimpleName(), tool);
    }
    if (opts.extraTools() != null) {
      for (Class<? extends BaseTool> tool : opts.extraTools()) {
        builtTools.put(tool.getSimpleName(), tool);
      }
    }
    this.tools = Collections.unmodifiableMap(builtTools);

    String serverUrl = opts.url() != null ? opts.url() : Config.SERVER_URL;

    List<Map<String, Object>> toolSchemas = ToolToSchemaConverter.convertAll(this.tools);

    this.client =
        new HttpClient(
            serverUrl,
            opts.model(),
            this.driver.platform(),
            toolSchemas,
            planner,
            excludeAttributes);

    this.model = client.model();
    LOG.info("Using model: {}/{}", this.model.provider().value(), this.model.name());

    this.cache = new Cache(this.client);
    LOG.info("Using HTTP client with server: {}", this.client.baseUrl());
  }

  // ---------------------------------------------------------------
  // Public accessors

  public BaseDriver driver() {
    return driver;
  }

  public HttpClient client() {
    return client;
  }

  public Cache cache() {
    return cache;
  }

  public Model model() {
    return model;
  }

  public Map<String, Class<? extends BaseTool>> tools() {
    return tools;
  }

  // ---------------------------------------------------------------
  // Actions

  /**
   * Executes a series of steps to achieve the given goal. Named {@code act} because {@code do} is a
   * reserved word in Java.
   */
  public DoResult act(String goal) {
    return executeDo(goal);
  }

  /** Alias for {@link #act(String)} - as close as possible to other clients {@code do}. */
  public DoResult do_(String goal) {
    return executeDo(goal);
  }

  /** Assert that the statement is true about the current view. */
  public String check(String statement) {
    return check(statement, new CheckOptions(false));
  }

  public String check(String statement, CheckOptions opts) {
    return Retry.execute(
        () -> {
          boolean vision = opts != null && opts.vision();
          driver.resetAccessibilityTree();
          HttpClient.RetrieveResult result =
              client.retrieve(
                  "Is the following true or false - " + statement,
                  driver.accessibilityTree().toStr(),
                  driver.title(),
                  driver.url(),
                  vision ? driver.screenshot() : null,
                  driver.app());
          if (!Boolean.TRUE.equals(result.result().boxedValue())) {
            throw new AssertionError(result.explanation());
          }
          return result.explanation();
        });
  }

  /** Extract data described in natural language from the current view. */
  public Object get(String data) {
    return get(data, new GetOptions(false));
  }

  public Object get(String data, GetOptions opts) {
    return Retry.execute(
        () -> {
          boolean vision = opts != null && opts.vision();
          driver.resetAccessibilityTree();
          HttpClient.RetrieveResult result =
              client.retrieve(
                  data,
                  driver.accessibilityTree().toStr(),
                  driver.title(),
                  driver.url(),
                  vision ? driver.screenshot() : null,
                  driver.app());
          Data value = result.result();
          if (value == null || value.isNoop()) {
            return result.explanation();
          }
          return value.toObject();
        });
  }

  /**
   * Finds an element in the accessibility tree and returns the native driver element.
   *
   * @param description Natural language description of the element to find.
   * @return Native driver element (Selenium WebElement, Playwright Locator, or Appium WebElement).
   */
  public Object find(String description) {
    return Retry.execute(
        () -> {
          driver.resetAccessibilityTree();
          FindElementResult response =
              client.findElement(description, driver.accessibilityTree().toStr(), driver.app());
          int id = response.id();
          return driver.findElement(id);
        });
  }

  /**
   * Creates an area for the agents to work within. This is useful for narrowing down the context or
   * focus of the agents' actions, checks and data retrievals.
   *
   * <p>Note that if the area cannot be found, the topmost area of the accessibility tree will be
   * used, which is equivalent to the whole page.
   *
   * @param description The description of the area.
   * @return An instance of the Area class that represents the area of the accessibility tree to
   *     use.
   */
  public Area area(String description) {
    driver.resetAccessibilityTree();
    FindElementResult response =
        client.findArea(description, driver.accessibilityTree().toStr(), driver.app());
    int id = response.id();
    return new Area(
        id,
        response.explanation(),
        driver,
        driver.accessibilityTree().scopeToArea(id),
        tools,
        client);
  }

  /**
   * Adds a new learning example on what steps should be take to achieve the goal.
   *
   * @param goal The goal to be achieved. Use same format as in {@link #act(String)}.
   * @param actions A list of actions to achieve the goal.
   */
  public void learn(String goal, List<String> actions) {
    client.addExample(goal, actions == null ? List.of() : List.copyOf(actions));
  }

  /** Clear all learning examples from the server session. */
  public void clearLearnExamples() {
    client.clearExamples();
  }

  /** Session statistics from the server. */
  public Map<String, Object> stats() {
    return client.stats();
  }

  /** Close the HTTP client/session and the driver. */
  public void quit() {
    try {
      client.quit();
    } finally {
      driver.quit();
    }
  }

  @Override
  public void close() {
    quit();
  }

  // ---------------------------------------------------------------
  // Internals

  private DoResult executeDo(String goal) {
    return Retry.execute(
        () -> {
          String app = driver.app();
          // Start from a fresh snapshot; it stays cached until the next step resets it.
          driver.resetAccessibilityTree();
          BaseAccessibilityTree initial = driver.accessibilityTree();
          String beforeTree = changeAnalysis ? initial.toStr() : null;
          String beforeUrl = changeAnalysis ? driver.url() : null;

          HttpClient.PlanResult plan = client.planActions(goal, initial.toStr(), app);
          String explanation = plan.explanation();

          List<DoStep> executedSteps = new ArrayList<>();
          List<String> steps = plan.steps();
          for (int idx = 0; idx < steps.size(); idx++) {
            String step = steps.get(idx);
            // Use the initial tree for the first step, a fresh tree afterwards.
            if (idx > 0) {
              driver.resetAccessibilityTree();
            }
            BaseAccessibilityTree tree = driver.accessibilityTree();
            HttpClient.ActionResult action = client.executeAction(goal, step, tree.toStr(), app);

            if (explanation.equals(goal)) {
              explanation = action.explanation();
            }

            List<String> calledTools = new ArrayList<>();
            for (DoStep toolCall : action.actions()) {
              calledTools.add(BaseTool.executeToolCall(toolCall, tools, driver));
            }
            executedSteps.add(new DoStep(step, calledTools));
          }

          String changes = "";
          if (changeAnalysis && !executedSteps.isEmpty()) {
            try {
              driver.resetAccessibilityTree();
              changes =
                  client.analyzeChanges(
                      beforeTree, beforeUrl, driver.accessibilityTree().toStr(), driver.url(), app);
            } catch (RuntimeException e) {
              LOG.warn("Error analyzing changes", e);
            }
          }

          return new DoResult(explanation, executedSteps, changes);
        });
  }

  private static BaseDriver wrapDriver(Object driver) {
    if (driver == null) {
      throw new IllegalArgumentException("driver must not be null");
    }
    // Order matters:
    //   - BaseDriver first, so already-wrapped drivers (returned by
    //     Alumni.driver()) pass through unchanged. This lets callers spin
    //     up a secondary Alumni bound to extra tools against the same
    //     underlying driver.
    //   - AppiumDriver before WebDriver, because AppiumDriver extends
    //     RemoteWebDriver (and thus WebDriver).
    return switch (driver) {
      case BaseDriver wrapped -> wrapped;
      case io.appium.java_client.AppiumDriver appium -> new AppiumDriver(appium);
      case com.microsoft.playwright.Page page -> new PlaywrightDriver(page);
      case org.openqa.selenium.WebDriver webDriver -> new SeleniumDriver(webDriver);
      default -> throw new UnsupportedOperationException("Driver " + driver + " not implemented");
    };
  }

  // ---------------------------------------------------------------
  // Option records

  /**
   * Construction-time options for {@link Alumni}. Any {@code null} field falls back to the
   * corresponding value from {@link Config} (i.e. the {@code ALUMNIUM_*} environment variables).
   */
  public record Options(
      String url,
      Model model,
      List<Class<? extends BaseTool>> extraTools,
      Boolean planner,
      Boolean changeAnalysis,
      Set<String> excludeAttributes) {
    public Options() {
      this(null, null, List.of(), null, null, null);
    }

    public Options {
      extraTools = extraTools == null ? List.of() : List.copyOf(extraTools);
      if (excludeAttributes != null) {
        excludeAttributes = Collections.unmodifiableSet(new LinkedHashSet<>(excludeAttributes));
      }
    }

    public Options withUrl(String url) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }

    public Options withModel(Model model) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }

    public Options withExtraTools(List<Class<? extends BaseTool>> extraTools) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }

    public Options withPlanner(Boolean planner) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }

    public Options withChangeAnalysis(Boolean changeAnalysis) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }

    public Options withExcludeAttributes(Set<String> excludeAttributes) {
      return new Options(url, model, extraTools, planner, changeAnalysis, excludeAttributes);
    }
  }

  /** Vision flag for {@link #get(String, GetOptions)}. */
  public record GetOptions(boolean vision) {
    public GetOptions() {
      this(false);
    }

    public GetOptions withVision(boolean vision) {
      return new GetOptions(vision);
    }
  }

  /** Vision flag for {@link #check(String, CheckOptions)}. */
  public record CheckOptions(boolean vision) {
    public CheckOptions() {
      this(false);
    }

    public CheckOptions withVision(boolean vision) {
      return new CheckOptions(vision);
    }
  }
}

File: packages/python/src/alumnium/alumni.py
from asyncio import AbstractEventLoop
from os import getenv

from appium.webdriver.webdriver import WebDriver as Appium
from playwright.async_api import Page as PageAsync
from playwright.sync_api import Page
from retry import retry
from selenium.webdriver.remote.webdriver import WebDriver

from . import CHANGE_ANALYSIS, DELAY, EXCLUDE_ATTRIBUTES, PLANNER, RETRIES
from .area import Area
from .cache import Cache
from .clients.http_client import HttpClient
from .clients.typecasting import Data
from .drivers import Element
from .drivers.appium_driver import AppiumDriver
from .drivers.playwright_async_driver import PlaywrightAsyncDriver
from .drivers.playwright_driver import PlaywrightDriver
from .drivers.selenium_driver import SeleniumDriver
from .logutils import get_logger
from .models import Model
from .result import DoResult, DoStep
from .tools import BaseTool

logger = get_logger(__name__)


class Alumni:
    def __init__(
        self,
        driver: Page | WebDriver | tuple[PageAsync, AbstractEventLoop],
        model: Model | None = None,
        extra_tools: list[type[BaseTool]] | None = None,
        url: str | None = None,
        planner: bool | None = None,
        change_analysis: bool | None = None,
        exclude_attributes: set[str] | None = None,
    ):
        planner = planner if planner is not None else PLANNER
        self.change_analysis = change_analysis if change_analysis is not None else CHANGE_ANALYSIS
        exclude_attributes = exclude_attributes if exclude_attributes is not None else EXCLUDE_ATTRIBUTES

        if isinstance(driver, Appium):
            self.driver = AppiumDriver(driver)
        elif isinstance(driver, Page):
            self.driver = PlaywrightDriver(driver)
        elif (
            isinstance(driver, tuple) and isinstance(driver[0], PageAsync) and isinstance(driver[1], AbstractEventLoop)
        ):
            # Asynchronous Playwright driver requires a shared event loop
            self.driver = PlaywrightAsyncDriver(driver[0], driver[1])
        elif isinstance(driver, WebDriver):
            self.driver = SeleniumDriver(driver)
        else:
            raise NotImplementedError(f"Driver {driver} not implemented")

        self.tools = {}
        for tool in self.driver.supported_tools | set(extra_tools or []):
            self.tools[tool.__name__] = tool

        server_url = url or getenv("ALUMNIUM_SERVER_URL")
        if server_url:
            logger.info(f"Using HTTP client with server: {server_url}")
        else:
            logger.info("Using HTTP client with auto-managed local server")

        self.client = HttpClient(
            server_url,
            model,
            self.driver.platform,
            self.tools,
            planner,
            exclude_attributes,
        )
        self.model = self.client.model

        logger.info(f"Using model: {self.model.provider.value}/{self.model.name}")

        self.cache = Cache(self.client)

    def quit(self) -> None:
        self.client.quit()
        self.driver.quit()

    @retry(tries=RETRIES, delay=DELAY, logger=logger)  # pyright: ignore[reportArgumentType]
    def do(self, goal: str) -> DoResult:
        """
        Executes a series of steps to achieve the given goal.

        Args:
            goal: The goal to be achieved.

        Returns:
            DoResult containing the explanation and executed steps with their actions.
        """
        app = self.driver.app
        self.driver.reset_accessibility_tree()
        initial_accessibility_tree = self.driver.accessibility_tree
        before_tree = initial_accessibility_tree.to_str() if self.change_analysis else None
        before_url = self.driver.url if self.change_analysis else None
        explanation, steps = self.client.plan_actions(goal, initial_accessibility_tree.to_str(), app=app)

        executed_steps = []
        for idx, step in enumerate(steps):
            # Use the initial tree for the first step, a fresh tree afterwards.
            if idx > 0:
                self.driver.reset_accessibility_tree()
            accessibility_tree = self.driver.accessibility_tree
            actor_explanation, actions = self.client.execute_action(goal, step, accessibility_tree.to_str(), app=app)

            # When planner is off, explanation is just the goal — replace with actor's reasoning.
            if explanation == goal:
                explanation = actor_explanation

            called_tools = []
            for tool_call in actions:
                called_tool = BaseTool.execute_tool_call(tool_call, self.tools, self.driver)
                called_tools.append(called_tool)

            executed_steps.append(DoStep(name=step, tools=called_tools))

        changes = ""
        if self.change_analysis and executed_steps:
            try:
                assert before_tree is not None
                assert before_url is not None
                self.driver.reset_accessibility_tree()
                changes = self.client.analyze_changes(
                    before_accessibility_tree=before_tree,
                    before_url=before_url,
                    after_accessibility_tree=self.driver.accessibility_tree.to_str(),
                    after_url=self.driver.url,
                )
            except Exception as e:
                logger.error(f"Error analyzing changes: {e}")

        return DoResult(explanation=explanation, steps=executed_steps, changes=changes)

    @retry(tries=RETRIES, delay=DELAY, logger=logger)  # pyright: ignore[reportArgumentType]
    def check(self, statement: str, vision: bool = False) -> str:
        """
        Checks a given statement true or false.

        Args:
            statement: The statement to be checked.
            vision: A flag indicating whether to use a vision-based verification via a screenshot. Defaults to False.

        Returns:
            The summary of verification result.

        Raises:
            AssertionError: If the verification fails.
        """
        self.driver.reset_accessibility_tree()
        explanation, value = self.client.retrieve(
            f"Is the following true or false - {statement}",
            self.driver.accessibility_tree.to_str(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
            app=self.driver.app,
        )
        assert value, explanation
        return explanation

    @retry(tries=RETRIES, delay=DELAY, logger=logger)  # pyright: ignore[reportArgumentType]
    def get(self, data: str, vision: bool = False) -> Data:
        """
        Extracts requested data from the page.

        Args:
            data: The data to extract.
            vision: A flag indicating whether to use a vision-based extraction via a screenshot. Defaults to False.

        Returns:
            The extracted data. If data cannot be extracted, returns the explanation string.
        """
        self.driver.reset_accessibility_tree()
        explanation, value = self.client.retrieve(
            data,
            self.driver.accessibility_tree.to_str(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
            app=self.driver.app,
        )
        return explanation if value is None else value

    @retry(tries=RETRIES, delay=DELAY, logger=logger)  # pyright: ignore[reportArgumentType]
    def find(self, description: str) -> Element:
        """
        Finds an element in the accessibility tree and returns the native driver element.

        Args:
            description: Natural language description of the element to find.

        Returns:
            Native driver element (Selenium WebElement, Playwright Locator, or Appium WebElement).
        """
        self.driver.reset_accessibility_tree()
        response = self.client.find_element(description, self.driver.accessibility_tree.to_str(), app=self.driver.app)
        return self.driver.find_element(response["id"])

    def area(self, description: str) -> Area:
        """
        Creates an area for the agents to work within.
        This is useful for narrowing down the context or focus of the agents' actions, checks and data retrievals.

        Note that if the area cannot be found, the topmost area of the accessibility tree will be used,
        which is equivalent to the whole page.

        Args:
            description: The description of the area.

        Returns:
            Area: An instance of the Area class that represents the area of the accessibility tree to use.
        """
        self.driver.reset_accessibility_tree()
        accessibility_tree = self.driver.accessibility_tree
        response = self.client.find_area(description, accessibility_tree.to_str(), app=self.driver.app)
        return Area(
            id=response["id"],
            description=response["explanation"],
            driver=self.driver,
            accessibility_tree=accessibility_tree.scope_to_area(response["id"]),
            tools=self.tools,
            client=self.client,
        )

    def learn(self, goal: str, actions: list[str]) -> None:
        """
        Adds a new learning example on what steps should be take to achieve the goal.

        Args:
            goal: The goal to be achieved. Use same format as in `do`.
            actions: A list of actions to achieve the goal.
        """
        self.client.add_example(goal, actions)

    def clear_learn_examples(self) -> None:
        """
        Clears the learn examples.
        """
        self.client.clear_examples()

    @property
    def stats(self) -> dict[str, dict[str, int]]:
        """
        Returns the stats of the session.
        """
        return self.client.stats
